### ISSUE TITLE
Hide implementation of Value

### ISSUE THREAD

**Issue by cdunn2001** [open-source-parsers/jsoncpp#417]
And I guess the iterators too. Then again, since this will be a major version bump, maybe we should drop the iterators, and later consider the ideas in #288.

This will help with #120 and #6. It will also let us use a "wiping allocator" as for #415 and #152.

**Comment by cdunn2001** [comment]
I've started an `impl` branch in my own fork. I'm not sure when I'll time to finish this. Needs several hours of work.

**Comment by cdunn2001** [comment]
New plan:
* Create `Json::Thing` as the new API.
* When we are ready,
  * Rename Thing->Value and Value->ValImpl.
  * Move the public classes.

This way, we can develop the new API in parallel with ongoing development.
